server.js

```diff
@@ -1,35 +1,180 @@
-
 // Defines the port the app will run on. Defaults to 8080, but can be overridden
 // when starting the server. Example command to overwrite PORT env variable value:
 // PORT=9000 npm start
+
+const Song = mongoose.model("Song", {
+    id: Number,
+    trackName: String,
+    artistName: String,
+    genre: String,
+    bpm: Number,
+    energy: Number,
+    danceability: Number,
+    loudness: Number,
+    liveness: Number,
+    valence: Number,
+    length: Number,
+    acousticness: Number,
+    speechiness: Number,
+    popularity: Number
+})
+if(process.env.RESET_DB){
+  const resetDataBase = async () => {
+    await Song.deleteMany();
+    topMusicData.forEach(singleSong => {
+      const newSong = new Song (singleSong);
+      newSong.save();
+    })
+
+  }
+  resetDataBase();
+}
 const port = process.env.PORT || 8080;
 const app = express();
 
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({
+      error: `Service unavailable`
+    })
+  }
+})
 // Add middlewares to enable cors and json body parsing
 app.use(cors());
 app.use(express.json());
 
 // Start defining your routes here
 app.get("/", (req, res) => {
-  res.send("Hello Technigo!");
+  res.send({ "Routes": listEndpoints(app)});
 });
 
+
+/// All songs
+app.get("/songs", async (req, res) => {
+  const allSongs = await Song.find({});
+    res.status(200).json({
+    success: true,
+    body: allSongs
+    });
+  });
+
+  //sorts all songs based on popularity rating
+  app.get("/songs/popularity", async (req, res) => {
+    try {
+      let popular = await Song.find()
+      popular = popular.sort((a, b) => b.popularity - a.popularity)
+      res.json({ length: popular.length, data: popular })
+    } catch {
+      res.status(400).json ({ error: `invalid request` })
+    }
+  })
+
+
+///endpoint that returns only one single item
+app.get("/id/:id", async (req, res) => {
+  const { id } = req.params;
+  try {
+    const singleSongID = await Song.findById(req.params.id);
+    
+    
+
+    if (singleSongID) {
+      res.status(200).json({
+        success: true,
+        body: singleSongID
+      });
+    } else {
+      res.status(404).json({
+        success: false,
+        status_code: 404,
+        body: {
+          message: `Could not find the song with provided ID ${id}`
+        }
+      });
+    }
+  } catch(error) {
+    res.status(400).json({
+      success: false,
+      status_code: 400,
+      body: {
+        message: `Invalid id ${id}`
+      }
+    });
+  }
+  
+});
+
+
+// app.get("/songs/genre/:genre/danceability/:danceability", async (req, res) => {
+app.get("/songs/", async (req, res) => {
+
+  const {genre, danceability} = req.query;
+  const response = {
+    success: true,
+    body: {}
+  }
+  const matchAllRegex = new RegExp(".*");
+  const matchAllNumeric = new RegExp("[0-9]");
+  const genreQuery = genre ? genre : {$regex: matchAllRegex,  $options: 'i' };
```

Comment:
this is different from Daniels code that looks like this ``` const genreQuery = genre ? genre : matchAllRegex;``` what are you thing to do? I dont understand.